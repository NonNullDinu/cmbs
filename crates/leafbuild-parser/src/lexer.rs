//! Contains the [`Lexer`] structure and its internals.
//! It is used as an external lexer to [`lalrpop`](https://github.com/lalrpop/lalrpop).
use crate::GrmError;
use leafbuild_ast::span::Span;
use logos::Logos;

/// The type of the token
#[derive(Logos, Copy, Clone, Debug, PartialEq, PartialOrd, Eq)]
pub enum Tk {
    #[token("+=")]
    PlusEq,
    #[token("-=")]
    MinusEq,
    #[token("*=")]
    MulEq,
    #[token("/=")]
    DivEq,
    #[token("%=")]
    ModEq,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Mul,
    #[token("/")]
    Slash,
    #[token("%")]
    Percent,
    #[token("==")]
    EqualEqual,
    #[token(">=")]
    GreaterEqual,
    #[token(">")]
    GreaterThan,
    #[token("<=")]
    LessEqual,
    #[token("<")]
    LessThan,
    #[token("!=")]
    NotEqual,
    #[token("=")]
    Equal,
    #[token("<<")]
    ShiftLeft,
    #[token(">>")]
    ShiftRight,
    #[token("(")]
    LParen,
    #[token("[")]
    LBracket,
    #[token("{")]
    LBrace,
    #[token(")")]
    RParen,
    #[token("]")]
    RBracket,
    #[token("}")]
    RBrace,
    #[token(".")]
    Dot,
    #[token(":")]
    Colon,
    #[token("?")]
    QMark,
    #[token(";")]
    Semicolon,
    #[token(",")]
    Comma,
    #[token("~")]
    Tilda,
    #[token("and")]
    And,
    #[token("or")]
    Or,
    #[token("not")]
    Not,
    #[token("in")]
    In,
    #[token("let")]
    Let,
    #[token("if")]
    If,
    #[token("else")]
    Else,
    #[token("foreach")]
    ForEach,
    #[token("continue")]
    Continue,
    #[token("break")]
    Break,
    #[token("return")]
    Return,
    #[token("true")]
    True,
    #[token("false")]
    False,
    #[token("fn")]
    Fn,
    #[regex("([1-9][0-9]*|0x[0-9a-fA-F]+|0b[01]+|0[0-7]+|0)[uU]?[lL]?")]
    Number,
    #[regex("[a-zA-Z_][a-zA-Z0-9_]*")]
    Id,
    #[regex(r#"'(\\['nt\\]|[^'\\])+'"#)]
    String,
    #[regex(r#"'''([^']*|'[^']|''[^'])*'''"#)]
    MultilineString,
    #[regex(r#"//[^\n]*"#, logos::skip)]
    SingleLineComment,
    #[regex(r"/\*([^*]|\**[^*/])*\*+/", logos::skip)]
    BlockComment,
    // #[token("\n")]
    // Newline,
    #[error]
    #[regex(r"[ \n\t\r]+", logos::skip)]
    // fix for https://github.com/maciejhirsz/logos/issues/180
    #[regex(r"/\*([^*]|\*+[^*/])*\*?")]
    Error,
}

/// The token structure
///
// Lalrpop throws tokens back if it cannot make sense of them, so this is part of the return type
// of [`crate::parse`], and so needs to be public.
#[derive(Debug, Clone, PartialOrd, PartialEq, Eq)]
pub struct Token<'data> {
    pub(crate) token: Tk,
    pub(crate) data: &'data str,
}
pub type LxrSpanned<Tok, Loc, Error> = Result<(Loc, Tok, Loc), Error>;

#[allow(missing_debug_implementations)]
pub struct Lexer<'a> {
    lexer: logos::Lexer<'a, Tk>,
}

impl<'a> Lexer<'a> {
    pub(crate) fn new(s: &'a str) -> Self {
        let lexer = Tk::lexer(s);
        Self { lexer }
    }
}

/// A lexical error.
/// Happens when the lexer wasn't able to understand some input
#[derive(Debug, Copy, Clone, PartialOrd, PartialEq, Eq)]
pub struct LexicalError {
    pub(crate) token: Tk,
    pub(crate) span: Span,
}

impl<'a> Iterator for Lexer<'a> {
    type Item = LxrSpanned<Token<'a>, usize, GrmError>;

    fn next(&mut self) -> Option<Self::Item> {
        self.lexer.next().map(|token| match token {
            Tk::Error => Err(GrmError::from(LexicalError {
                token,
                span: Span::from(self.lexer.span()),
            })),
            token => {
                let span = self.lexer.span();
                Ok((
                    span.start,
                    Token {
                        token,
                        data: self.lexer.slice(),
                    },
                    span.end,
                ))
            }
        })
    }
}

#[cfg(test)]
mod tests;
