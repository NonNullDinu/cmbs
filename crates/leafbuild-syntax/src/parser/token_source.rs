use crate::syntax_kind::SyntaxKind;

pub trait TokenSource {
    fn current(&self) -> Token;
    fn lookahead(&self, n: usize) -> Token;
    fn bump(&mut self);
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Token {
    pub syntax_kind: SyntaxKind,
}
