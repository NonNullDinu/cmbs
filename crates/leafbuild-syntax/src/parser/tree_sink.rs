use crate::syntax_kind::SyntaxKind;

use super::error::ParseError;
pub trait TreeSink {
    fn token(&mut self, kind: SyntaxKind);
    fn start_node(&mut self, kind: SyntaxKind);
    fn finish_node(&mut self);
    fn error(&mut self, error: ParseError);
}
