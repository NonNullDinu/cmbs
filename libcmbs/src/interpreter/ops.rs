use crate::interpreter::{Value, ValueTypeMarker};

pub(crate) trait OpPlus<T, Result>
    where T: ValueTypeMarker, Result: ValueTypeMarker {
    fn plus(val: Value<T>) -> Value<Result>;
}