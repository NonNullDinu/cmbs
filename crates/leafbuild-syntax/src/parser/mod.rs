//! # The parser code
//! Also see [the syntax reference](https://leafbuild.github.io/leafbuild/dev/syntax_ref.html).

use std::fmt;
use std::ops::Range;

use rowan::{TextRange, TextSize};
use token_source::token_source_from_lexer::LexerWrap;

use leafbuild_stdx::CopyTo;

use crate::lexer::Lexer;
use crate::SyntaxKind::{self, *};
use crate::T;

use self::text_tree_sink::TextTreeSink;
use self::token_source::{FindProperty, ForwardToken, Token};
use self::tree_sink::TreeSink;
use self::{
    error::ParseError,
    event::Event,
    marker::{CompletedMarker, Marker},
    token_source::TokenSource,
};

pub mod error;
mod event;
mod marker;
pub mod text_tree_sink;
pub mod token_source;
pub mod tree_sink;

trait IsTrivia: Copy {
    fn is_trivia(self) -> bool;
}

impl IsTrivia for SyntaxKind {
    fn is_trivia(self) -> bool {
        self == Self::WHITESPACE || self == Self::SINGLE_LINE_COMMENT || self == Self::BLOCK_COMMENT
    }
}

/// A span in the source code
#[derive(Copy, Clone, Default, Eq, PartialEq, Hash)]
#[repr(transparent)]
pub struct Span {
    /// The underlying text range
    pub text_range: TextRange,
}

impl From<Range<TextSize>> for Span {
    fn from(range: Range<TextSize>) -> Self {
        Self {
            text_range: TextRange::new(range.start, range.end),
        }
    }
}

impl From<Range<u32>> for Span {
    fn from(range: Range<u32>) -> Self {
        Self {
            text_range: TextRange::new(range.start.into(), range.end.into()),
        }
    }
}

impl fmt::Debug for Span {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.text_range.fmt(f)
    }
}

pub(crate) struct Parser<'ts> {
    /// tokens
    source: &'ts mut dyn TokenSource,
    events: Vec<Event>,
}

#[allow(clippy::inline_always)]
impl<'ts> Parser<'ts> {
    fn parse(&mut self) {
        let marker = self.start();
        loop {
            if self.at(EOF) {
                break;
            }
            parse_lang_item(self);
        }

        marker.complete(self, ROOT);
    }

    #[inline(always)]
    fn at(&self, kind: SyntaxKind) -> bool {
        self.current() == kind
    }

    #[inline(always)]
    fn at_any<const N: usize>(&self, kinds: [SyntaxKind; N]) -> bool {
        let current = self.current();
        std::array::IntoIter::new(kinds).any(|it| it == current)
    }

    #[inline(always)]
    fn bump(&mut self, kind: SyntaxKind) {
        assert!(self.eat(kind));
    }

    #[inline(always)]
    fn bump_any(&mut self) {
        let kind = self.current();
        if kind == EOF {
            return;
        }

        self.do_bump(kind, 1);
    }

    fn do_bump(&mut self, kind: SyntaxKind, n_raw_tokens: u8) {
        for _ in 0..n_raw_tokens {
            self.source.bump();
        }
        self.push_event(Event::Token { kind, n_raw_tokens })
    }

    fn expect(&mut self, kind: SyntaxKind) -> bool {
        if self.eat(kind) {
            return true;
        }
        self.error(format!("expected {:?}, got {:?}", kind, self.current()));
        false
    }

    pub(crate) fn error<E: Into<ParseError>>(&mut self, msg: E) {
        let msg = msg.into();
        self.push_event(Event::Error { msg })
    }

    #[inline(always)]
    fn eat(&mut self, kind: SyntaxKind) -> bool {
        if self.at(kind) {
            self.do_bump(kind, 1);
            true
        } else {
            false
        }
    }

    #[inline(always)]
    fn current(&self) -> SyntaxKind {
        self.nth(0)
    }

    #[inline(always)]
    fn nth(&self, n: usize) -> SyntaxKind {
        self.source.lookahead(n).syntax_kind
    }

    #[inline(always)]
    fn require_newline(&mut self) {
        self.expect(NEWLINE);
    }

    #[inline(always)]
    fn has_newline(&self) -> bool {
        self.at(NEWLINE)
    }

    #[inline(always)]
    fn bump_to(&mut self, forward_token: ForwardToken) {
        let bumps = self.source.bump_to(forward_token);
        for _ in 0..bumps {
            self.bump_any();
        }
    }

    #[inline(always)]
    fn next_not_newline(&self) -> ForwardToken {
        self.source.find(FindProperty::KindIsNot(NEWLINE))
    }

    #[inline(always)]
    fn bump_to_if_next_non_newline_is(&mut self, kind: SyntaxKind) -> bool {
        let mut tk = ForwardToken::default();
        let k = self.next_not_newline().copy_to(&mut tk).kind;
        if k == kind {
            self.bump_to(tk);
            true
        } else {
            false
        }
    }

    #[inline(always)]
    fn bump_to_if_next_non_newline_is_any<const KINDS_SIZE: usize>(
        &mut self,
        kinds: [SyntaxKind; KINDS_SIZE],
    ) -> bool {
        let mut tk = ForwardToken::default();
        let k = self.next_not_newline().copy_to(&mut tk).kind;
        if std::array::IntoIter::new(kinds).any(|it| it == k) {
            self.bump_to(tk);
            true
        } else {
            false
        }
    }

    #[inline(always)]
    fn skip_newlines(&mut self) {
        while self.at(NEWLINE) {
            self.do_bump(NEWLINE, 1);
        }
    }

    #[inline(always)]
    fn unexpected(&mut self) {
        let current = self.current();
        if current != EOF {
            self.do_bump(current, 1);
        }

        self.error(format!("Unexpected {:?}", current));
    }

    #[inline(always)]
    fn start(&mut self) -> Marker {
        let pos = self.events.len() as u32;
        self.push_event(Event::tombstone());
        Marker::new(pos)
    }

    #[inline(always)]
    fn push_event(&mut self, event: Event) {
        self.events.push(event)
    }
}
///
pub fn parse(text: &str) -> (rowan::GreenNode, Vec<(ParseError, TextSize)>) {
    let lexer = Lexer::new(text);
    let tokens: Vec<_> = lexer
        .map(|(kind, span)| Token {
            syntax_kind: kind,
            len: span.text_range.len(),
        })
        .collect();
    let mut lexer = LexerWrap::new(&tokens);
    let mut sink = TextTreeSink::new(text, &tokens);

    parse_to_sink(&mut lexer, &mut sink);
    sink.finish()
}

/// parses `text`
pub fn parse_to_sink(source: &mut dyn TokenSource, sink: &mut dyn TreeSink) {
    let mut p = Parser {
        source,
        events: vec![],
    };
    p.parse();
    event::process(sink, p.events);
}

#[inline]
fn parse_lang_item(p: &mut Parser) {
    p.skip_newlines();
    match p.current() {
        T![struct] => parse_struct_decl(p),
        _ => parse_statement(p),
    }
}

fn parse_statement(p: &mut Parser) {
    let tok = p.current();
    match tok {
        T!['(']
        | T!['[']
        | T!['{']
        | T![+]
        | T![-]
        | T![not]
        | T![true]
        | T![false]
        | T![NumLit]
        | T![Id]
        | T![Str]
        | T![MultilineStr]
        | T![if] => {
            let assingment_marker = p.start();
            parse_expr(p);
            // test assignment
            // a = b

            // test unit_to_unit_assignment
            // () = ()
            if p.at_any([T![+=], T![-=], T![*=], T![/=], T![%=], T![=]]) {
                // consume the `=` / `+=` / ...
                p.bump_any();

                // test assignment_expr_on_next_line
                // x =
                //     1

                parse_expr(p);

                p.require_newline();
                assingment_marker.complete(p, Assignment);
            } else {
                // test freestanding_expr
                // 1

                p.require_newline();
                assingment_marker.abandon(p);
            }
        }
        T![let] => parse_declaration(p),
        T![foreach] => parse_foreach(p),
        T![continue] | T![break] | T![return] => parse_control_stmt(p),
        _ => {
            p.unexpected();
        }
    }
}

fn parse_expr(p: &mut Parser) {
    p.skip_newlines();

    // test precedence_parsing
    // x = 1 + 2 * 3 % - 4 ( 5 )

    let marker = p.start();
    parse_precedence_8_expr(p);
    marker.complete(p, Expr);
}

fn parse_tuple_expr(p: &mut Parser) {
    p.skip_newlines();
    assert!(is_tuple_expr_start(p));

    parse_tt(p, TupleExpr, T!['('], Some(T![,]), T![')'], parse_expr)
}

fn is_tuple_expr_start(p: &mut Parser) -> bool {
    p.at(T!['('])
}

fn parse_array_expr(p: &mut Parser) {
    p.skip_newlines();
    assert!(is_array_expr_start(p));

    parse_tt(p, ArrayLitExpr, T!['['], Some(T![,]), T![']'], parse_expr)
}

fn is_array_expr_start(p: &mut Parser) -> bool {
    p.at(T!['['])
}

fn parse_primary(p: &mut Parser) -> CompletedMarker {
    p.skip_newlines();
    let marker = p.start();

    if is_array_expr_start(p) {
        parse_array_expr(p)
    } else if is_tuple_expr_start(p) {
        parse_tuple_expr(p)
    } else if is_conditional_start(p) {
        // test if_condition_in_expr
        // a = if b {} else {}

        // test if_condition_in_expr_stretched
        // a = if b
        // {
        //
        // }
        //
        // else
        // {
        //
        //
        // }
        parse_conditional(p)
    } else if is_expr_block_start(p) {
        parse_expr_block(p)
    } else if p.at_any([T![NumLit], T![Id]]) {
        p.bump_any();
    } else if is_string_lit(p) {
        parse_string(p)
    } else {
        p.unexpected();
    }
    marker.complete(p, PrimaryExpr)
}

fn is_string_lit(p: &mut Parser) -> bool {
    p.at_any([T![Str], T![MultilineStr]])
}

fn parse_string(p: &mut Parser) {
    assert!(is_string_lit(p));
    p.bump_any()
}

fn parse_tt(
    p: &mut Parser,
    outer_kind: SyntaxKind,
    start_tok: SyntaxKind,
    separator: Option<SyntaxKind>,
    end_tok: SyntaxKind,
    mut f: impl FnMut(&mut Parser),
) {
    let marker = p.start();
    if !p.expect(start_tok) {
        marker.abandon(p);
        return;
    }

    p.skip_newlines();
    while !p.at_any([EOF, end_tok]) {
        f(p);

        if let Some(separator) = separator {
            p.skip_newlines();
            if p.eat(separator) {
            } else if !p.at(end_tok) {
                p.error(format!(
                    "expected {:?} or {:?}, got {:?}",
                    separator,
                    end_tok,
                    p.current()
                ));
            }
        }
        p.skip_newlines();
    }

    p.expect(end_tok);
    marker.complete(p, outer_kind);
}

fn parse_precedence_1_expr(p: &mut Parser) -> CompletedMarker {
    p.skip_newlines();
    let mut marker = parse_primary(p);

    // test index_on_second_line_is_array_lit_expr
    // x = a
    // [1]

    while p.at_any([T!['('], T!['[']]) {
        if p.at(T!['(']) {
            let new_marker = marker.precede(p);
            marker = parse_f_call(p, new_marker);
        } else if p.at(T!['[']) {
            let new_marker = marker.precede(p);
            marker = parse_index_expr(p, new_marker);
        }
    }

    marker
}

fn parse_f_call(p: &mut Parser, marker: Marker) -> CompletedMarker {
    // test simple_function_call
    // x = f(1, 2, a = b)

    // test function_call
    // x = f(1, 2, a = b)

    // test named_args_only_function_call
    // x = f(a = b, c = d)

    // test err func_call_on_second_line
    // x = f
    // (1, 2, a = b)

    parse_tt(p, FnCallArgsList, T!['('], Some(T![,]), T![')'], parse_farg);
    marker.complete(p, FnCallExpr)
}

fn parse_farg(p: &mut Parser) {
    p.skip_newlines();
    if is_kexpr_start(p) {
        parse_kexpr(p)
    } else {
        parse_expr(p)
    }
}

fn parse_kexpr(p: &mut Parser) {
    p.skip_newlines();
    assert!(is_kexpr_start(p));
    let marker = p.start();

    p.bump(T![Id]);
    p.bump(T![=]);

    parse_expr(p);

    marker.complete(p, KExpr);
}

fn is_kexpr_start(p: &mut Parser) -> bool {
    p.at(T![Id]) && p.nth(1) == T![=]
}

fn parse_index_expr(p: &mut Parser, marker: Marker) -> CompletedMarker {
    let brackets_marker = p.start();
    p.bump(T!['[']);
    parse_expr(p); // expr
    p.expect(T![']']);
    brackets_marker.complete(p, IndexExprBrackets);
    marker.complete(p, IndexExpr)
}

fn parse_precedence_2_expr(p: &mut Parser) -> CompletedMarker {
    p.skip_newlines();
    if p.at_any([T![+], T![-]]) {
        let marker = p.start();
        p.bump_any();
        parse_precedence_2_expr(p);
        marker.complete(p, PrefixUnaryOpExpr)
    } else {
        parse_precedence_1_expr(p)
    }
}

fn parse_infix_binop<const N: usize>(
    p: &mut Parser,
    ops: [SyntaxKind; N],
    mut lower: impl FnMut(&mut Parser) -> CompletedMarker,
) -> CompletedMarker {
    p.skip_newlines();
    let mut completed = lower(p);

    // test expr_with_binary_infix_operators_on_same_line_and_second_operand_on_second_line
    // x = 1 +
    // 2 +
    // f(
    //      4
    // )
    // y = 3 *
    // 6 %
    // 78

    // test expr_with_binary_infix_operators_on_next_line
    // x = 1
    // + 2
    // + f(
    //      4
    // )
    // y = 3
    // * 6
    // % 78

    while p.bump_to_if_next_non_newline_is_any(ops) {
        let prec = completed.precede(p);
        p.bump_any();
        lower(p);

        completed = prec.complete(p, InfixBinOpExpr);
    }

    completed
}

fn parse_precedence_3_expr(p: &mut Parser) -> CompletedMarker {
    parse_infix_binop(p, [T![*], T![/], T![%]], parse_precedence_2_expr)
}

fn parse_precedence_4_expr(p: &mut Parser) -> CompletedMarker {
    parse_infix_binop(p, [T![+], T![-]], parse_precedence_3_expr)
}

fn parse_precedence_5_expr(p: &mut Parser) -> CompletedMarker {
    parse_infix_binop(p, [T![<], T![<=], T![>], T![>=]], parse_precedence_4_expr)
}

fn parse_precedence_6_expr(p: &mut Parser) -> CompletedMarker {
    parse_infix_binop(p, [T![==], T![!=]], parse_precedence_5_expr)
}

fn parse_precedence_7_expr(p: &mut Parser) -> CompletedMarker {
    parse_infix_binop(p, [T![and]], parse_precedence_6_expr)
}

fn parse_precedence_8_expr(p: &mut Parser) -> CompletedMarker {
    parse_infix_binop(p, [T![or]], parse_precedence_7_expr)
}

fn parse_expr_block(p: &mut Parser) {
    p.skip_newlines();
    parse_tt(p, ExprBlock, T!['{'], None, T!['}'], parse_statement)
}

fn is_expr_block_start(p: &mut Parser) -> bool {
    p.at(T!['{'])
}

fn parse_declaration(p: &mut Parser) {
    // test declaration
    // let a = b

    // test var_declaration_with_proper_expr_as_value
    // let x = 1
    p.skip_newlines();
    let mk = p.start();
    p.expect(T![let]);
    p.expect(T![Id]);

    p.expect(T![=]);

    parse_expr(p);

    p.require_newline();

    mk.complete(p, Declaration);
}

fn is_conditional_start(p: &mut Parser) -> bool {
    p.at(T![if])
}

fn parse_conditional(p: &mut Parser) {
    p.skip_newlines();

    // test if_else_condition
    // if a {} else {}

    // test if_else_if_else_condition
    // if a {} else if b {} else {}

    assert!(is_conditional_start(p));

    let marker = p.start();
    parse_conditional_branch(p);
    // test if_else_if_else_condition_stretched
    // if ()
    //
    // {
    //
    // }
    //
    // else if
    // ()
    //
    // {
    //
    // }
    //
    // else {
    //
    // }

    // test if_condition_on_next_line
    // if
    //      1
    // {}
    while p.bump_to_if_next_non_newline_is(T![else]) {
        p.bump(T![else]);
        if p.bump_to_if_next_non_newline_is(T![if]) {
            parse_conditional_branch(p);
        } else {
            parse_expr_block(p);
            break;
        }
    }

    marker.complete(p, Conditional);
}

fn parse_conditional_branch(p: &mut Parser) {
    p.skip_newlines();
    let marker = p.start();

    // consume the IF_KW
    p.bump(T![if]);

    parse_expr(p);

    parse_expr_block(p);
    marker.complete(p, ConditionalBranch);
}

fn parse_foreach(p: &mut Parser) {
    p.skip_newlines();

    // test foreach_basic
    // foreach a in b {}

    let marker = p.start();

    p.bump(T![foreach]);
    parse_expr(p);
    p.expect(T![in]);
    parse_expr(p);
    parse_expr_block(p);

    marker.complete(p, Foreach);

    p.require_newline();
}

fn is_control_stmt(p: &mut Parser) -> bool {
    p.at_any([T![continue], T![break], T![return]])
}

fn parse_control_stmt(p: &mut Parser) {
    // test continue_test
    // continue

    // test break_without_value
    // break

    // test break_with_value
    // break 1

    // test break_with_value_on_new_line
    // break
    //        1

    // test return_without_value
    // return

    // test return_with_value
    // return 1

    // test return_with_value_on_new_line
    // return
    //         1
    p.skip_newlines();
    assert!(is_control_stmt(p));

    let marker = p.start();
    if p.eat(T![continue]) {
    } else if (p.eat(T![return]) || p.eat(T![break])) && !p.at(NEWLINE) {
        parse_expr(p);
    }
    p.require_newline();

    marker.complete(p, ControlStatement);
}

fn parse_struct_decl(p: &mut Parser) {
    p.skip_newlines();

    // test empty_struct
    // struct test {}

    // test struct_with_one_field
    // struct test {
    //     i: i32
    // }

    // test struct_with_field_and_generics
    // struct test {
    //     i: i32<a, b>
    // }

    // test err struct_with_field_name_but_no_type
    // struct test {
    //     i
    // }

    let marker = p.start();
    p.bump(T![struct]);
    p.expect(T![Id]);
    p.skip_newlines();
    parse_tt(
        p,
        StructFieldList,
        T!['{'],
        None,
        T!['}'],
        parse_struct_field,
    );
    marker.complete(p, StructDecl);
    p.require_newline();
}

fn parse_struct_field(p: &mut Parser) {
    let marker = p.start();
    p.expect(T![Id]);
    p.expect(T![:]);
    parse_type_ref(p);
    p.require_newline();
    marker.complete(p, StructField);
}

fn parse_type_ref(p: &mut Parser) {
    let marker = p.start();
    p.expect(T![Id]);
    if p.at(T![<]) {
        parse_type_ref_generics(p);
    }
    marker.complete(p, TypeRef);
}

fn parse_type_ref_generics(p: &mut Parser) {
    parse_tt(
        p,
        TypeRefGenerics,
        T![<],
        Some(T![,]),
        T![>],
        parse_type_ref,
    )
}
