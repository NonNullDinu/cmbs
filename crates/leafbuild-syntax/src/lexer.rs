// @generated by xtask generate-grammar
use crate::parser::Span;
use crate::syntax_kind::SyntaxKind;
use crate::T;
use leafbuild_stdx::Let;
use logos::Logos;
use std::convert::TryInto;
use std::ops::Range;
#[derive(Logos, Copy, Clone, Debug, PartialEq, PartialOrd, Eq)]
pub enum Tk {
    #[regex(r#"[a-zA-Z_][a-zA-Z0-9_]*"#)]
    Id,
    #[regex(r#"([1-9][0-9]*|0x[0-9a-fA-F]+|0b[01]+|0[0-7]+|0)[uU]?[lL]?"#)]
    NumLit,
    #[regex(r#"'(\\['nt\\]|[^'\\])+'"#)]
    Str,
    #[regex(r#"'''([^']*|'[^']|''[^'])*'''"#)]
    MultilineStr,
    #[regex(r#"//[^\n]*"#)]
    SingleLineComment,
    #[regex(r#"/\*([^*]|\**[^*/])*\*+/"#)]
    BlockComment,
    #[regex(r#"[ \t\r]+"#)]
    Whitespace,
    #[token("+=")]
    PlusEq,
    #[token("-=")]
    MinusEq,
    #[token("*=")]
    MulEq,
    #[token("/=")]
    DivEq,
    #[token("%=")]
    ModEq,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Asterisk,
    #[token("/")]
    Slash,
    #[token("%")]
    Percent,
    #[token("==")]
    EqEq,
    #[token(">=")]
    GreaterEq,
    #[token(">")]
    Greater,
    #[token("<=")]
    LessEq,
    #[token("<")]
    Less,
    #[token("=")]
    Eq,
    #[token("!=")]
    NotEq,
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token(".")]
    Dot,
    #[token(":")]
    Colon,
    #[token("?")]
    QMark,
    #[token(";")]
    Semicolon,
    #[token(",")]
    Comma,
    #[token("and")]
    AndKw,
    #[token("or")]
    OrKw,
    #[token("not")]
    NotKw,
    #[token("in")]
    InKw,
    #[token("let")]
    LetKw,
    #[token("if")]
    IfKw,
    #[token("else")]
    ElseKw,
    #[token("foreach")]
    ForeachKw,
    #[token("continue")]
    ContinueKw,
    #[token("break")]
    BreakKw,
    #[token("return")]
    ReturnKw,
    #[token("true")]
    TrueKw,
    #[token("false")]
    FalseKw,
    #[token("fn")]
    FnKw,
    #[token("struct")]
    StructKw,
    #[token("\n")]
    Newline,
    #[error]
    #[regex(r"/\*([^*]|\*+[^*/])*\*?")]
    Error,
}
impl From<Tk> for SyntaxKind {
    fn from(tk: Tk) -> Self {
        use SyntaxKind::*;
        match tk {
            Tk::Id => T![Id],
            Tk::NumLit => T![NumLit],
            Tk::Str => T![Str],
            Tk::MultilineStr => T![MultilineStr],
            Tk::SingleLineComment => T![SingleLineComment],
            Tk::BlockComment => T![BlockComment],
            Tk::Whitespace => T![Whitespace],
            Tk::PlusEq => T ! [+=],
            Tk::MinusEq => T ! [-=],
            Tk::MulEq => T ! [*=],
            Tk::DivEq => T ! [/=],
            Tk::ModEq => T ! [%=],
            Tk::Plus => T ! [+],
            Tk::Minus => T ! [-],
            Tk::Asterisk => T ! [*],
            Tk::Slash => T ! [/],
            Tk::Percent => T ! [%],
            Tk::EqEq => T ! [==],
            Tk::GreaterEq => T ! [>=],
            Tk::Greater => T ! [>],
            Tk::LessEq => T ! [<=],
            Tk::Less => T ! [<],
            Tk::Eq => T ! [=],
            Tk::NotEq => T ! [!=],
            Tk::LParen => T!['('],
            Tk::RParen => T![')'],
            Tk::LBracket => T!['['],
            Tk::RBracket => T![']'],
            Tk::LBrace => T!['{'],
            Tk::RBrace => T!['}'],
            Tk::Dot => T ! [.],
            Tk::Colon => T ! [:],
            Tk::QMark => T ! [?],
            Tk::Semicolon => T ! [;],
            Tk::Comma => T ! [,],
            Tk::AndKw => T![and],
            Tk::OrKw => T![or],
            Tk::NotKw => T![not],
            Tk::InKw => T![in],
            Tk::LetKw => T![let],
            Tk::IfKw => T![if],
            Tk::ElseKw => T![else],
            Tk::ForeachKw => T![foreach],
            Tk::ContinueKw => T![continue],
            Tk::BreakKw => T![break],
            Tk::ReturnKw => T![return],
            Tk::TrueKw => T![true],
            Tk::FalseKw => T![false],
            Tk::FnKw => T![fn],
            Tk::StructKw => T![struct],
            Tk::Newline => T!['\n'],
            Tk::Error => ERROR,
        }
    }
}
#[allow(missing_debug_implementations)]
pub struct Lexer<'a> {
    lexer: logos::SpannedIter<'a, Tk>,
}
impl<'a> Lexer<'a> {
    pub(crate) fn new(s: &'a str) -> Self {
        let lexer = Tk::lexer(s).spanned();
        Self { lexer }
    }
}
impl<'a> Iterator for Lexer<'a> {
    type Item = (SyntaxKind, Span);
    fn next(&mut self) -> Option<Self::Item> {
        self.lexer.next().map(|(token, span)| {
            (
                token.into(),
                span.let_(|it| -> Range<u32> {
                    it.start.try_into().unwrap()..it.end.try_into().unwrap()
                })
                .into(),
            )
        })
    }
}
