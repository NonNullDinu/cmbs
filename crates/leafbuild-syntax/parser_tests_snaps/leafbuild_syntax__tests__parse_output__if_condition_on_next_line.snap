---
source: crates/leafbuild-syntax/src/parser/mod.rs
expression: "if\n     1\n{}\n"

---
- BUILD_DEFINITION 0..13
  - EXPR_EVAL_STMT 0..13
    - PRIMARY_EXPR 0..12
      - CONDITIONAL 0..12
        - "if" IF_KW 0..2
        - "\n" NEWLINE 2..3
        - "     " WHITESPACE 3..8
        - PRIMARY_EXPR 8..9
          - "1" INT_NUMBER 8..9
        - "\n" NEWLINE 9..10
        - BLOCK 10..12
          - "{" L_CURLY 10..11
          - "}" R_CURLY 11..12
    - "\n" NEWLINE 12..13

